### PR TITLE
Generator for Pronto with Gitlab CI

### PR DESCRIPTION
**Pull request by coolprobn**
Tasks
---

- Generator for Pronto with Gitlab CI
- Tests
- Idempotent Generator (Doesn't add duplicate changes no matter how many time you run it)
- New Helper file:
  - for searching ruby version from Gemfile and .ruby-version file
  - for checking Gemfile for the existing gem and installing only if it doesn't exist

### CODE REVIEW

**Comment by coolprobn** [comment]
@abhaynikam Could you review this PR when you have time?

Thanks.

File: lib/generators/boring/pronto/base_generator.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        enum: %w[brakeman flay reek rubocop],
+        default: []
+
+      include BoringGenerators::GeneratorHelper
+
+      def add_pronto_gem
+        say "Adding pronto gem", :green
+
+        Bundler.with_unbundled_env do
+          check_and_install_gem "pronto"
```
Comment (abhaynikam) [inline lib/generators/boring/pronto/base_generator.rb:23]: it doesn't seem this run install from code. does it?

File: lib/generators/boring/pronto/gitlab_ci/install/install_generator.rb [hunk 2]
```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+
+require "generators/boring/pronto/base_generator"
+require "boring_generators/generator_helper"
+
+module Boring
+  module Pronto
+    module GitlabCi
+      class InstallGenerator < Boring::Pronto::BaseGenerator
+        desc "Adds pronto gem with various extensions and configure for Gitlab CI"
```
Comment (abhaynikam) [inline lib/generators/boring/pronto/gitlab_ci/install/install_generator.rb:10]:
```suggestion
        desc "Adds Pronto gem with various extensions and configures it for GitLab CI."
```

File: lib/generators/boring/pronto/gitlab_ci/install/install_generator.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+        def add_configuration
+          @ruby_version = options.ruby_version || app_ruby_version
+
+          if File.exists?(".gitlab-ci.yml")
+            add_configuration_to_existing_file
+            add_lint_stage_to_existing_file
+            show_readme
+          else
+            say "We couldn't find .gitlab-ci.yml in your project, creating new file with Pronto configurations ...",
```
Comment (abhaynikam) [inline lib/generators/boring/pronto/gitlab_ci/install/install_generator.rb:25]:
```suggestion
            say "Creating .gitlab-ci.yml with Pronto configurations.",
```

File: lib/generators/boring/pronto/gitlab_ci/install/install_generator.rb [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+          end
+        end
+
+        private
+
+        def add_configuration_to_existing_file
+          say "Adding Pronto configurations to .gitlab-ci.yml", :green
+
+          if gitlab_ci_file_content["pronto"]
+            say "Pronto configurations already exist! Skipping them ...",
```
Comment (abhaynikam) [inline lib/generators/boring/pronto/gitlab_ci/install/install_generator.rb:38]:
```suggestion
            say "Skipping Pronto configurations",
```

File: lib/generators/boring/pronto/gitlab_ci/install/install_generator.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        class_option :ruby_version, type: :string, aliases: "-rv"
+
+        include BoringGenerators::GeneratorHelper
+
+        def add_configuration
+          @ruby_version = options.ruby_version || app_ruby_version
+
+          if File.exists?(".gitlab-ci.yml")
+            add_configuration_to_existing_file
+            add_lint_stage_to_existing_file
```
Comment (abhaynikam) [inline lib/generators/boring/pronto/gitlab_ci/install/install_generator.rb:22]: Why is this needed? Won't `add_configuration_to_existing_file` add lint and ruby verson configuration?

**Review by abhaynikam — CHANGES_REQUESTED** [review]
Left some comments

File: lib/generators/boring/pronto/base_generator.rb (hunk 1 above)
Comment (coolprobn) [reply lib/generators/boring/pronto/base_generator.rb:23]: We have created a new helper method that checks the Gemfile and return early if it finds the gem. Else it will just run the gem command provided by the rails generator.

```
    def check_and_install_gem(*args)
      gem_name, = args

      gem_file_content_array = File.readlines("Gemfile")

      gem_exists = gem_file_content_array.any? { |line| line.include?(gem_name) }

      if gem_exists
        say "#{gem_name} is already in the Gemfile, skipping it ...", :yellow
      else
        gem *args
      end
    end
```

File: lib/generators/boring/pronto/gitlab_ci/install/install_generator.rb (hunk 5 above)
Comment (coolprobn) [reply lib/generators/boring/pronto/gitlab_ci/install/install_generator.rb:22]: They are a bit different.

- add_configuration_to_existing_file adds `pronto` block and configure setups required to run Pronto
- add_lint_stage_to_existing_file is adds "lint" to "stages" so that linting is now enabled in Gitlab CI

**Comment by coolprobn** [comment]
@abhaynikam Resolved all comments.

**Comment by abhaynikam** [comment]
@coolprobn can you rebese your branch?

**Comment by coolprobn** [comment]
@abhaynikam Done!

**Review by abhaynikam — APPROVED** [review]
